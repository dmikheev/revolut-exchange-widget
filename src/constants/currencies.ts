export enum Currency {
    USD = 'USD',
    EUR = 'EUR',
    GBP = 'GBP',
}

export const SymbolByCurrency = {
    [Currency.USD]: '$',
    [Currency.EUR]: '€',
    [Currency.GBP]: '£',
};
