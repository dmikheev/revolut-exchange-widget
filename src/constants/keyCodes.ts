export enum KeyCode {
  ENTER = 13,
  DIGIT_1 = 49,
}
