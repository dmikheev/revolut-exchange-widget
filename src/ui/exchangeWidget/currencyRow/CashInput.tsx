import React from 'react';
import { TextFieldProps } from "@material-ui/core/TextField";
import { TextField } from "@material-ui/core";
import { isCashStringValid } from "../../../utils/isCashStringValid";

const CashInput: React.FC<TextFieldProps> = ({ onChange: onChangeProp, ...restProps }) => {
  const onChange = React.useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      if (!onChangeProp) {
        return;
      }

      if (!isCashStringValid(event.target.value)) {
        return;
      }

      onChangeProp(event);
    },
    [onChangeProp],
  );

  return (
    <TextField {...restProps} onChange={onChange}/>
  );
};
export default CashInput;
