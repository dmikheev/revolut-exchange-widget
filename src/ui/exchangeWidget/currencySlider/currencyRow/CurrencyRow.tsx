import {
  createStyles,
  TextField,
  Theme,
  Typography,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import { TextFieldProps } from '@material-ui/core/TextField';
import classNames from 'classnames';
import React from 'react';
import { Currency } from '../../../../constants/currencies';

import styles from './CurrencyRow.module.css';

export interface ICurrencyRowProps {
  className?: string;
  balance: number;
  currency: Currency;
  isRateDisabled?: boolean;
}
const CurrencyRow: React.FC<ICurrencyRowProps> = ({
  className,
  balance,
  currency,
}) => {
  return (
    <div className={classNames(styles.wrap, className)}>
      <div className={styles.column}>
        <Typography variant="h3">
          {currency}
        </Typography>
        <Typography variant="body2" className={styles.note}>
          You have {balance} {currency}
        </Typography>
      </div>
      <div className={styles.column}>
        <StyledTextField/>
      </div>
    </div>
  );
};
export default CurrencyRow;

const textFieldStyles = (theme: Theme) => createStyles({
  input: theme.typography.h3,
});

type IStyledTextFieldComponentProps = TextFieldProps & WithStyles<typeof textFieldStyles>;
const StyledTextFieldComponent: React.FC<IStyledTextFieldComponentProps> =
  ({ classes, ...restProps }) => {
    const inputProps = {
      classes: {
        root: classes.input,
      },
    };

    return (
      <TextField {...restProps} InputProps={inputProps}/>
    );
  };
const StyledTextField = withStyles(textFieldStyles)(StyledTextFieldComponent);
