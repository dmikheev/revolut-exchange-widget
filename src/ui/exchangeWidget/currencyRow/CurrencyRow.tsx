import {
  CircularProgress,
  createStyles,
  FormControl,
  InputLabel,
  MenuItem,
  Select,
  Theme,
  Typography,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import { TextFieldProps } from '@material-ui/core/TextField';
import classNames from 'classnames';
import React from 'react';
import { Currency, SymbolByCurrency } from '../../../constants/currencies';
import { KeyCode } from '../../../constants/keyCodes';
import { cashFormat } from '../../../utils/cashFormat';
import CashInput from './CashInput';

import styles from './CurrencyRow.module.css';

interface ICurrencyRowProps {
  className?: string;
  amountStr: string;
  balance: number;
  currencies: Currency[];
  currency: Currency;
  sourceCurrency: Currency;
  isBalanceError?: boolean;
  isRateFetching?: boolean;
  isSourceCurrency?: boolean;
  rate?: number;
  onAmountChange(amount: string): void;
  onCurrencyChange(currency: Currency): void;
  onInputEnterPress?(): void;
}

const CurrencyRow: React.FC<ICurrencyRowProps> = ({
  className,
  amountStr,
  balance,
  currencies,
  currency,
  sourceCurrency,
  isBalanceError,
  isRateFetching,
  isSourceCurrency,
  rate,
  onAmountChange,
  onCurrencyChange,
  onInputEnterPress,
}) => {
  const onCurrencySelectChange = React.useCallback(
    (event: React.ChangeEvent<HTMLSelectElement>) => onCurrencyChange(event.target.value as Currency),
    [onCurrencyChange],
  );
  const onInputChange = React.useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => onAmountChange(event.target.value),
    [onAmountChange],
  );
  const onInputKeyPress = React.useCallback(
    (event: React.KeyboardEvent<HTMLInputElement>) => {
      if (onInputEnterPress && event.which === KeyCode.ENTER) {
        onInputEnterPress();
      }
    },
    [onInputEnterPress],
  );

  const selectItemsHtml = currencies.map((cur) => (
    <MenuItem key={cur} value={cur}>{cur}</MenuItem>
  ));

  let amountColumnHtml;
  if (isSourceCurrency || currency !== sourceCurrency) {
    let rateContentHtml;
    if (!isSourceCurrency) {
      const rateValueHtml = rate !== undefined && (
        <Typography variant="body2" className={styles.note}>
          {SymbolByCurrency[currency]}1 = {SymbolByCurrency[sourceCurrency]}{cashFormat(rate)}
        </Typography>
      );
      const rateProgressHtml = isRateFetching && (
        <CircularProgress className={styles.rate_progress} size="0.8em" color="secondary"/>
      );

      rateContentHtml = (
        <React.Fragment>
          {rateValueHtml}
          {rateProgressHtml}
        </React.Fragment>
      );
    }

    amountColumnHtml = (
      <React.Fragment>
        <StyledTextField
          error={isBalanceError}
          value={amountStr}
          InputProps={{inputProps: {className: styles.input}}}
          onChange={onInputChange}
          onKeyPress={onInputKeyPress}
        />
        <div className={styles.rate_wrap}>
          {rateContentHtml}
        </div>
      </React.Fragment>
    );
  }

  return (
    <div className={classNames(styles.wrap, className)}>
      <div className={styles.column}>
        <FormControl>
          <InputLabel>{isSourceCurrency ? 'From' : 'To'}</InputLabel>
          <Select value={currency} onChange={onCurrencySelectChange}>
            {selectItemsHtml}
          </Select>
        </FormControl>
        <Typography variant="body2" className={styles.note} color={isBalanceError ? 'error' : undefined}>
          You have {SymbolByCurrency[currency]}{cashFormat(balance)}
        </Typography>
      </div>
      <div className={styles.column}>
        {amountColumnHtml}
      </div>
    </div>
  );
};
export default CurrencyRow;

const textFieldStyles = (theme: Theme) => createStyles({
  input: theme.typography.h3,
});

type IStyledTextFieldComponentProps = TextFieldProps & WithStyles<typeof textFieldStyles>;
const StyledTextFieldComponent: React.FC<IStyledTextFieldComponentProps> =
  ({classes, ...restProps}) => {
    const inputProps = {
      ...restProps.InputProps,
      classes: {
        root: classes.input,
      },
    };

    return (
      <CashInput {...restProps} InputProps={inputProps}/>
    );
  };
const StyledTextField = withStyles(textFieldStyles)(StyledTextFieldComponent);
