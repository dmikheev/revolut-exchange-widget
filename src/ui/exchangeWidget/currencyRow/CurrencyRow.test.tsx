import { CircularProgress, Select, Typography } from '@material-ui/core';
import { cleanup, render } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Currency, SymbolByCurrency } from '../../../constants/currencies';
import { KeyCode } from '../../../constants/keyCodes';
import CashInput from './CashInput';
import CurrencyRow, { ICurrencyRowProps, StyledTextField } from './CurrencyRow';

const currencies = [Currency.USD, Currency.EUR, Currency.GBP];
const nop = () => {};
const testProps: ICurrencyRowProps = {
  amountStr: '11',
  balance: 222,
  currencies,
  currency: Currency.EUR,
  rate: 3.3,
  sourceCurrency: Currency.USD,
  onAmountChange: nop,
  onCurrencyChange: nop,
};

afterEach(cleanup);

it('renders correctly', () => {
  const wrapper = shallow(<CurrencyRow {...testProps} className="test-class-name"/>);
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('doesn\'t show exchange rate for source currency', () => {
  const currencyFrom = Currency.USD;
  const currencyTo = Currency.EUR;
  const rate = 2;
  const rateText = `${SymbolByCurrency[currencyTo]}1 = ${SymbolByCurrency[currencyFrom]}${rate}`;
  const { queryByText } = render(
    <CurrencyRow
      {...testProps}
      currency={currencyTo}
      sourceCurrency={currencyFrom}
      rate={rate}
      isSourceCurrency={true}
    />
  );

  expect(queryByText(rateText)).toBeFalsy();
});

it('doesn\'t show the amount input if current currency equals source currency', () => {
  const currencyFrom = Currency.USD;
  const currencyTo = currencyFrom;
  const wrapper = mount(
    <CurrencyRow
      {...testProps}
      currency={currencyTo}
      sourceCurrency={currencyFrom}
    />
  );

  const input = wrapper.find(StyledTextField);
  expect(input).toHaveLength(0);
});

it('doesn\'t show exchange rate if current currency equals source currency', () => {
  const currencyFrom = Currency.USD;
  const currencyTo = currencyFrom;
  const rate = 1;
  const rateText = `${SymbolByCurrency[currencyTo]}1 = ${SymbolByCurrency[currencyFrom]}${rate}`;
  const { queryByText } = render(
    <CurrencyRow
      {...testProps}
      currency={currencyTo}
      sourceCurrency={currencyFrom}
      rate={rate}
    />
  );

  expect(queryByText(rateText)).toBeFalsy();
});

it('shows input with error state and balance text with error color if there is a balance error', () => {
  const wrapper = mount(
    <CurrencyRow
      {...testProps}
      amountStr="10"
      balance={100}
      isBalanceError={true}
    />
  );

  const inputFind = wrapper.find(CashInput);
  expect(inputFind).toHaveLength(1);
  expect(inputFind.props().error).toBeTruthy();

  const typographyFind = wrapper.find(Typography);
  const errorColorTypography = typographyFind.filter({ color: 'error' });
  expect(errorColorTypography).toHaveLength(1);
  expect(errorColorTypography.text()).toEqual(expect.stringContaining('You have'));
});

it('shows exchange rate loader if required props are specified and rate is fetching', () => {
  const wrapper = shallow(
    <CurrencyRow
      {...testProps}
      currency={Currency.EUR}
      sourceCurrency={Currency.USD}
      rate={2}
      isRateFetching={true}
    />
  );

  expect(wrapper.find(CircularProgress)).toHaveLength(1);
});

it('calls onAmountChange with the correct value when the input changes', () => {
  const value = '12.34';
  const onAmountChange = jest.fn();
  const wrapper = shallow(<CurrencyRow {...testProps} onAmountChange={onAmountChange}/>);

  const input = wrapper.find(StyledTextField);
  expect(input).toHaveLength(1);

  expect(input.props().onChange).toBeTruthy();
  input.props().onChange!({ target: { value } } as React.ChangeEvent<HTMLInputElement>);
  expect(onAmountChange).toBeCalledTimes(1);
  expect(onAmountChange).toHaveBeenLastCalledWith(value);
});

it('calls onInputEnterPress when Enter has been pressed in the input', () => {
  const onInputEnterPress = jest.fn();
  const wrapper = shallow(<CurrencyRow {...testProps} onInputEnterPress={onInputEnterPress}/>);

  const input = wrapper.find(StyledTextField);
  expect(input).toHaveLength(1);

  expect(input.props().onKeyPress).toBeTruthy();
  input.props().onKeyPress!({ which: KeyCode.DIGIT_1 } as React.KeyboardEvent<HTMLInputElement>);
  expect(onInputEnterPress).not.toBeCalled();

  input.props().onKeyPress!({ which: KeyCode.ENTER } as React.KeyboardEvent<HTMLInputElement>);
  expect(onInputEnterPress).toBeCalledTimes(1);
});

it('calls onCurrencyChange when currency has been selected', async () => {
  const currencyToChoose = Currency.EUR;
  const onCurrencyChange = jest.fn();
  const wrapper = shallow(
    <CurrencyRow
      {...testProps}
      currencies={[Currency.USD, Currency.EUR]}
      currency={Currency.USD}
      onCurrencyChange={onCurrencyChange}
    />
  );

  const select = wrapper.find(Select);
  expect(select).toHaveLength(1);
  expect(select.props().onChange).toBeTruthy();
  select.props().onChange!({ target: { value: currencyToChoose } } as React.ChangeEvent<HTMLSelectElement>, undefined);

  expect(onCurrencyChange).toBeCalledTimes(1);
  expect(onCurrencyChange).toHaveBeenLastCalledWith(currencyToChoose);
});
