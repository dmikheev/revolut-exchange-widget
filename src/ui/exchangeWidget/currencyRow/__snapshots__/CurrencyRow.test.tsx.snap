// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="wrap test-class-name"
>
  <div
    className="column"
  >
    <WithStyles(FormControl)>
      <WithStyles(WithFormControlContext(InputLabel))>
        To
      </WithStyles(WithFormControlContext(InputLabel))>
      <WithStyles(WithFormControlContext(Select))
        onChange={[Function]}
        value="EUR"
      >
        <WithStyles(MenuItem)
          key="USD"
          value="USD"
        >
          USD
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          key="EUR"
          value="EUR"
        >
          EUR
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          key="GBP"
          value="GBP"
        >
          GBP
        </WithStyles(MenuItem)>
      </WithStyles(WithFormControlContext(Select))>
    </WithStyles(FormControl)>
    <WithStyles(Typography)
      className="note"
      variant="body2"
    >
      You have 
      €
      222
    </WithStyles(Typography)>
  </div>
  <div
    className="column"
  >
    <WithStyles(StyledTextFieldComponent)
      InputProps={
        Object {
          "inputProps": Object {
            "className": "input",
            "inputMode": "numeric",
          },
        }
      }
      onChange={[Function]}
      onKeyPress={[Function]}
      value="11"
    />
    <div
      className="rate_wrap"
    >
      <WithStyles(Typography)
        className="note"
        variant="body2"
      >
        €
        1 = 
        $
        3.3
      </WithStyles(Typography)>
    </div>
  </div>
</div>
`;
